I pulled the 10.9 sources, applied this PR as a patch and then built sage.  The build succeeded.  But I get a traceback when I try to create a free group of rank 3 with `F = FreeGroup(3)`.  It doesn't look like an issue with root paths but it doesn't work either. I'd say it looks more like an issue with string encoding, or maybe with passing an empty group as an empty tuple..

Anyway, here is my traceback:

```
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 10.9, Release Date: 2026-05-04                    │
│ Using Python 3.14.3. Type "help()" for help.                       │
└────────────────────────────────────────────────────────────────────┘
sage: F = FreeGroup(3)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /var/tmp/sage-10.9/local/lib/python3.14/site-packages/sage/misc/cachefunc.pyx:1040, in sage.misc.cachefunc.CachedFunction.__call__()
   1039 try:
-> 1040     return self.cache[k]
   1041 except TypeError:  # k is not hashable

File sage/misc/weak_dict.pyx:708, in sage.misc.weak_dict.WeakValueDictionary.__getitem__()
--> 708             raise KeyError(k)

KeyError: ((<class 'sage.groups.free_group.FreeGroup_class'>, ('x0', 'x1', 'x2')), ())

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 F = FreeGroup(Integer(3))

File sage/misc/lazy_import.pyx:413, in sage.misc.lazy_import.LazyImport.__call__()
--> 413         return self.get_object()(*args, **kwds)

File /var/tmp/sage-10.9/local/lib/python3.14/site-packages/sage/groups/free_group.py:700, in FreeGroup(n, names, index_set, abelian, **kwds)
    698     from sage.groups.indexed_free_group import IndexedFreeGroup
    699     return IndexedFreeGroup(index_set, names=names, **kwds)
--> 700 return FreeGroup_class(names, **kwds)

File sage/misc/classcall_metaclass.pyx:320, in sage.misc.classcall_metaclass.ClasscallMetaclass.__call__()
--> 320             return cls.classcall(cls, *args, **kwds)

File sage/misc/cachefunc.pyx:1045, in sage.misc.cachefunc.CachedFunction.__call__()
-> 1045             w = self.f(*args, **kwds)

File /var/tmp/sage-10.9/local/lib/python3.14/site-packages/sage/structure/unique_representation.py:1195, in CachedRepresentation.__classcall__(cls, *args, **options)
   1180 @weak_cached_function(cache=128)  # automatically a staticmethod
   1181 def __classcall__(cls, *args, **options):
   1182     """
   1183     Construct a new object of this class or reuse an existing one.
   1184 
   (...)   1193         True
   1194     """
-> 1195     return super().__classcall__(cls, *args, **options)

File /var/tmp/sage-10.9/local/lib/python3.14/site-packages/sage/structure/unique_representation.py:598, in WithPicklingByInitArgs.__classcall__(cls, *args, **options)
    574 @staticmethod
    575 def __classcall__(cls, *args, **options):
    576     """
    577     Construct a new object of this class and store the arguments passed to ``__init__``.
    578 
   (...)    596         False
    597     """
--> 598     instance = typecall(cls, *args, **options)
    599     assert isinstance(instance, cls)
    600     if instance.__class__.__reduce__ == WithPicklingByInitArgs.__reduce__:

File sage/misc/classcall_metaclass.pyx:471, in sage.misc.classcall_metaclass.typecall()
--> 471     return (<PyTypeObject*>type).tp_call(cls, args, kwds)

File /var/tmp/sage-10.9/local/lib/python3.14/site-packages/sage/groups/free_group.py:741, in FreeGroup_class.__init__(self, generator_names, gap_group)
    720 """
    721 Python constructor.
    722 
   (...)    738     ('a', 'b')
    739 """
    740 if gap_group is None:
--> 741     gap_group = libgap.FreeGroup(generator_names)
    742 ParentLibGAP.__init__(self, gap_group)
    743 if not generator_names:

File sage/libs/gap/libgap.pyx:788, in sage.libs.gap.libgap.Gap.__getattr__()
--> 788             g = self.eval(name)

File sage/libs/gap/libgap.pyx:502, in sage.libs.gap.libgap.Gap.eval()
--> 502         initialize()

File sage/libs/gap/util.pyx:339, in sage.libs.gap.util.initialize()
--> 339     s = str_to_bytes(gap_roots, FS_ENCODING, "surrogateescape")

File sage/cpython/string.pxd:75, in sage.cpython.string.str_to_bytes()
---> 75     return _str_to_bytes(s, encoding, errors)

TypeError: expected str, list found
sage: 
```

--- a/src/sage/interfaces/gap.py
+++ b/src/sage/interfaces/gap.py
@@ -206,7 +206,7 @@ import pexpect
 
 import sage.interfaces.abc
 from sage.cpython.string import bytes_to_str
-from sage.env import GAP_ROOT_PATHS, SAGE_EXTCODE, SAGE_GAP_COMMAND, SAGE_GAP_MEMORY
+from sage.env import SAGE_EXTCODE, SAGE_GAP_COMMAND, SAGE_GAP_MEMORY
 from sage.interfaces.expect import (
     Expect,
     ExpectElement,
@@ -228,7 +228,7 @@ if SAGE_GAP_COMMAND is None:
     # Passing -A allows us to use a minimal GAP installation without
     # producing errors at start-up. The files sage.g and sage.gaprc are
     # used to load any additional packages that may be available.
-    gap_cmd = f'gap -A -l "{GAP_ROOT_PATHS}"'
+    gap_cmd = "gap -A"
     if SAGE_GAP_MEMORY is not None:
         gap_cmd += " -s " + SAGE_GAP_MEMORY + " -o " + SAGE_GAP_MEMORY
 else:

--- a/src/sage/interfaces/gap_workspace.py
+++ b/src/sage/interfaces/gap_workspace.py
@@ -17,7 +17,7 @@ import os
 import time
 import hashlib
 import subprocess
-from sage.env import DOT_SAGE, HOSTNAME, GAP_ROOT_PATHS
+from sage.env import DOT_SAGE, HOSTNAME
 
 
 def gap_workspace_file(system='gap', name='workspace', dir=None):
@@ -60,15 +60,9 @@ def gap_workspace_file(system='gap', name='workspace', dir=None):
     if dir is None:
         dir = os.path.join(DOT_SAGE, 'gap')
 
-    data = f'{GAP_ROOT_PATHS}'
-    for path in GAP_ROOT_PATHS.split(";"):
-        if not path:
-            # If GAP_ROOT_PATHS begins or ends with a semicolon,
-            # we'll get one empty path.
-            continue
-        sysinfo = os.path.join(path, "sysinfo.gap")
-        if os.path.exists(sysinfo):
-            data += subprocess.getoutput(f'. "{sysinfo}" && echo ":$GAP_VERSION:$GAParch"')
+    from sage.libs.gap.util import kernel_info
+    version, arch, paths = kernel_info()
+    data = ";".join(paths + [version, arch])
     h = hashlib.sha1(data.encode('utf-8')).hexdigest()
     return os.path.join(dir, f'{system}-{name}-{HOSTNAME}-{h}')
 

--- a/src/sage/libs/gap/util.pyx
+++ b/src/sage/libs/gap/util.pyx
@@ -36,6 +36,60 @@ from sage.interfaces.gap_workspace import prepare_workspace_dir
 ############################################################################
 
 
+def kernel_info():
+    r"""
+    Return the GAP version, architecture, and root paths in a tuple.
+
+    The first two are used as cache keys to invalidate old workspaces
+    in :func:`sage.interfaces.gap_workspace.gap_workspace_file`. The
+    root paths are used (required) to initialize libgap. In the past
+    we computed the root paths at build-time, but that may not work if
+    (say) the build and target hosts have different libdirs.
+
+    This is fast enough that it should suffice until a more reliable
+    method is available (GAP Github issue 6252).
+
+    OUTPUT:
+
+    A tuple with three elements:
+
+    1. The GAP version (str)
+    2. The GAP architecture (str)
+    3. A list of GAP's RootPaths (list of str)
+
+    TESTS:
+
+    Confirm the claims of our ``OUTPUT`` block::
+
+        sage: from sage.libs.gap.util import kernel_info
+        sage: ki = kernel_info()
+        sage: isinstance(ki, tuple)
+        True
+        sage: len(ki) == 3
+        True
+        sage: isinstance(ki[0], str)
+        True
+        sage: isinstance(ki[1], str)
+        True
+        sage: isinstance(ki[2], list)
+        True
+        sage: all( isinstance(p, str) for p in ki[2] )
+        True
+
+    """
+    import subprocess
+    from importlib.resources import files
+    systemfile = files("sage").joinpath("ext_data/gap/kernel-info.g")
+    cmd_and_args = ["gap", "--systemfile", systemfile]
+    result = subprocess.run(cmd_and_args,
+                            capture_output=True,
+                            check=True,
+                            text=True)
+    version, arch, roots = result.stdout.strip().split("\n")
+    roots = roots.split(";")
+    return (version, arch, roots)
+
+
 cdef class ObjWrapper():
     """
     Wrapper for GAP master pointers.
@@ -232,7 +286,8 @@ cdef initialize():
     argv[0] = "sage"
     argv[1] = "-A"
     argv[2] = "-l"
-    s = str_to_bytes(sage.env.GAP_ROOT_PATHS, FS_ENCODING, "surrogateescape")
+    gap_roots = kernel_info()[2]
+    s = str_to_bytes(";".join(gap_roots), FS_ENCODING, "surrogateescape")
     argv[3] = s
 
     argv[4] = "-m"

